protocols/gossipsub/src/types.rs

```diff
@@ -514,3 +586,233 @@ impl fmt::Display for PeerKind {
         f.write_str(self.as_ref())
     }
 }
+
+/// `RpcOut` sender that is priority aware.
+#[derive(Debug, Clone)]
+pub(crate) struct RpcSender {
+    cap: usize,
+    len: Arc<AtomicUsize>,
+    pub(crate) priority_sender: Sender<RpcOut>,
+    pub(crate) non_priority_sender: Sender<RpcOut>,
+    priority_receiver: Receiver<RpcOut>,
+    non_priority_receiver: Receiver<RpcOut>,
+}
+
+impl RpcSender {
+    /// Create a RpcSender.
+    pub(crate) fn new(cap: usize) -> RpcSender {
+        let (priority_sender, priority_receiver) = async_channel::unbounded();
```

Comment:
So IIUC the `priority_sender` will not send `PUBLISH` messages if its `queue_len` is above `cap`, but it will send `GRAFT`, `PRUNE`, `SUBSCRIBE` and `UNSUBSCRIBE` anyway in an unbounded way?

Reply:
yeah exactly!

Reply:
Following the principle to [bound everything](https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#bound-everything): it would be great to bound the channel even for `GRAFT`, etc. messages. IIUC this would mean creating a new priority class. There would be 1. low priority, 2. normal priority (`PUBLISH`) and 3. high priority (`GRAFT`, etc.).

Not sure it would fit in the scope of this PR, but switching to using configurable priority queues, where the number of priority classes can be custom, could make sense.

Reply:
Alright, we can keep it like this, and we'll revisit later if the priority queue grows too large.